message_ix/core.py

```diff
@@ -26,6 +27,111 @@ def __init__(self, mp, model, scenario=None, version=None, annotation=None,
         super().__init__(mp, model, scenario, version, scheme, annotation,
                          cache)
 
+    @lru_cache()
+    def _year_idx(self, name):
+        """Return (idx_set, idx_name) for 'year'-indexed dims of *name*."""
+        # NB Since item dimensionality is fixed, the result can be cached for
+        #    performance
+        return list(filter(
+            lambda e: e[0] == 'year',
+            zip(self.idx_sets(name), self.idx_names(name))
+        ))
+
+    def _year_as_int(self, name, df):
+        """Convert 'year'-indexed columns of *df* to int dtypes."""
+        year_idx = self._year_idx(name)
+        if len(year_idx):
```

Comment:
Just a question - will it work in case we later add values to dataframe? How will new year values be handled - as integers (expected) or default (category or inherited dtype)?

Reply:
You're talking about something like:
```
df = scen.par('parname')
df = df.append(rows)
df.loc[:, 'year'] = 9999
```

When `rows` has 'year' values, they will be handled as `int`. So this means that they (or the literal `9999`) are not restricted to a limited set of categorical values, e.g. the contents of `scen.set('year')`.

However, if one then tries to `scen.add_par('parname', df)`, then it will fail.

I think this is the current behaviour, i.e. one can construct any kind of DataFrame, but ixmp/message_ix does not complain about it until one adds it.

Does this seem adequate?

Reply:
My original observation (or surprise) was about different treating empty and non-empry dataframes.